core/event_loop.go

```diff
@@ -23,25 +23,42 @@ import (
 // EventLoopWithInterrupt starts an instance of the agent created with NewAgent().
 // Agent is stopped when <closeChan> is closed, a user interrupt (SIGINT), or a
 // terminate signal (SIGTERM) is received.
-func EventLoopWithInterrupt(agent *Agent, closeChan chan struct{}) error {
-	err := agent.Start()
-	if err != nil {
-		agent.Error("Error loading core: ", err)
-		return err
-	}
-
-	sigChan := make(chan os.Signal, 1)
-	signal.Notify(sigChan, os.Interrupt)
-	signal.Notify(sigChan, syscall.SIGTERM)
-	select {
-	case <-sigChan:
-		agent.Println("Interrupt received, returning.")
-	case <-closeChan:
-	}
+func EventLoopWithInterrupt(agent *Agent, closeCh chan struct{}) error {
+	_, errCh := Run(agent, closeCh)
+	return <-errCh
+}
 
-	err = agent.Stop()
-	if err != nil {
-		agent.Errorf("Agent stop error '%+v'", err)
-	}
-	return err
+// Run starts an instance of the agent created with NewAgent().
+// Agent is stopped when <closeChan> is closed, , a user interrupt (SIGINT), or a
+// terminate signal (SIGTERM) is received.
+//
+// Returns readyCh which will be closed when the agent is ready - warning readyCh may never be closed if the agent fails on start
+// Returns errCh which will receive all errors associated with starting or stopping the event loop
+//    errCh will be closed when the agent is successfully stopped
+func Run(agent *Agent, closeCh chan struct{}) (<-chan struct{}, <-chan error) {
+	// We need two slots in the errCh, one for errors on Start(), one for errors on Stop()
+	errCh := make(chan error, 2)
```

Comment:
There isn't any situation that `errCh` would need capacity of **2**.

* If the **start of agent fails**, error is sent on `errCh` channel and goroutine returns, which then closes the channel in defer, so no more errors will be sent.
* If the **start is ok** (nothing sent to `errCh` channel) and **stop of agent fails**, error is sent on `errCh` channel and goroutine ends after that, which then also closes the channel in defer, so again no more errors will be sent.